torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,244 @@
+                qconfig_dict
+
+    Return:
+        None, module is modified inplace with qconfig attached
+    """
+    if not hasattr(module, 'qconfig'):
+        module.qconfig = None
+        if qconfig_dict and prefix in qconfig_dict:
+            module.qconfig = qconfig_dict[prefix]
+        else:
+            module.qconfig = qconfig_parent
+
+    for name, child in module.named_children():
+        module_prefix = prefix + '.' + name if prefix else name
+        propagate_qconfig_helper(child, qconfig_dict, module.qconfig, module_prefix)
+
+def propagate_qconfig(module, qconfig_dict=None):
+    r"""Propagate qconfig through the module hierarchy and assign `qconfig`
+    attribute on each leaf module
+
+    Args:
+        module: input module
+        qconfig_dict: dictionary that maps from name of submodule to quantization
+            configuration, qconfig applies to all submodules of a given
+            module unless qconfig for the submodules are specified(when the
+            submodule already has qconfig attribute)
+
+    Return:
+        None, module is modified inplace with qconfig attached
+    """
+    propagate_qconfig_helper(module, qconfig_dict)
+
+def _observer_forward_hook(self, input, output):
+    r"""Forward hook that calls observer on the output
+    """
+    self.observer(output)
+
+def add_observer(module):
+    r"""Add observer for the leaf child of the module.
+
+    This function insert observer module to all leaf child module that
+    has a valid qconfig attribute.
+
+    Args:
+        module: input module with qconfig attributes for all the leaf modules
+        that we want to quantize
+
+    Return:
+        None, module is modified inplace with added observer modules and
+            forward_hooks
+    """
+    for child in module.children():
+        add_observer(child)
+
+    # Insert observers only for leaf nodes, note that this observer is for
+    # the output of the module, for input QuantStub will observe them
+    if hasattr(module, 'qconfig') and module.qconfig is not None and len(module._modules) == 0:
+        # observer and hook will be gone after we swap the module
+        module.add_module('observer', module.qconfig.activation())
+        module.register_forward_hook(_observer_forward_hook)
+
+class QuantWrapper(nn.Module):
+    r"""A wrapper class that wraps the input module, adds QuantStub and
+    DeQuantStub and surround the call to module with call to quant and dequant
+    modules.
+
+    This is used by the `quantization` utility functions to add the quant and
+    dequant modules, before `convert` function `QuantStub` will just be observer,
+    it observes the input tensor, after `convert`, `QuantStub`
+    will be swapped to `nnq.Quantize` which does actual quantization. Similarly
+    for `DeQuantStub`.
+    """
+    def __init__(self, module):
+        super(QuantWrapper, self).__init__()
+        assert hasattr(module, 'qconfig'), 'Please add qconfig to module before \
```

Comment:
Wait, why is that? you can just make QuantWrapper insertion **before** propagate_qconfig and everything should work as long as qwrappers are on the leaf nodes.

Probably best is to just check whether qconfig is present on module and then copy it, otherwise keep it empty